File: src/main/java/net/thedragonteam/armorplus/worldgen/nbt/StructureGenNBT.java
/*
 * Copyright (c) TheDragonTeam 2016-2017.
 */

package net.thedragonteam.armorplus.worldgen.nbt;

import net.minecraft.server.MinecraftServer;
import net.minecraft.util.ResourceLocation;
import net.minecraft.util.Rotation;
import net.minecraft.util.math.BlockPos;
import net.minecraft.world.World;
import net.minecraft.world.WorldServer;
import net.minecraft.world.biome.Biome;
import net.minecraft.world.chunk.IChunkProvider;
import net.minecraft.world.gen.IChunkGenerator;
import net.minecraft.world.gen.structure.template.PlacementSettings;
import net.minecraft.world.gen.structure.template.Template;
import net.minecraftforge.fml.common.IWorldGenerator;
import net.thedragonteam.thedragonlib.util.LogHelper;

import java.util.Random;

import static net.minecraftforge.common.BiomeDictionary.*;
import static net.thedragonteam.armorplus.DevUtils.enableTowerDevEnv;
import static net.thedragonteam.armorplus.ModConfig.DebugConfig.debugMode;
import static net.thedragonteam.armorplus.ModConfig.WorldGenConfig.tower;
import static net.thedragonteam.armorplus.util.Utils.setRL;
import static net.thedragonteam.armorplus.util.WorldGenUtils.*;

/**
 * @author Sokratis Fotkatzikis - TheDragonTeam
 **/
public class StructureGenNBT implements IWorldGenerator {

    public static final ResourceLocation TOWER = setRL("tower");

    @Override
    public void generate(Random random, int chunkX, int chunkZ, World world, IChunkGenerator chunkGenerator, IChunkProvider chunkProvider) {

        int posX = x16(chunkX) + random.nextInt(CHUNK_SIZE);
        int posZ = x16(chunkZ) + random.nextInt(CHUNK_SIZE);

        BlockPos posXZ = new BlockPos(posX, 1, posZ);

        Biome biome = world.getBiomeForCoordsBody(posXZ);

        if (!(world instanceof WorldServer)) return;

        WorldServer serverworld = (WorldServer) world;

        int posY = getGroundFromAbove(world, posX, posZ);
        BlockPos basePos = new BlockPos(posX, posY, posZ);

        if (tower.shouldOnlyGenerateInTheOverworld) {
            if (world.provider.getDimension() == 0) {
                getBiomes(Type.HOT).forEach(hotBiome -> {
                    boolean isBiomeEligible = (!hasType(hotBiome, Type.SAVANNA) && !hasType(hotBiome, Type.JUNGLE));
                    if (isBiomeEligible && biome == hotBiome) {
                        this.generateTower(serverworld, random, basePos);
                    }
                });
            }
        } else {
            getBiomes(Type.HOT).forEach(hotBiome -> {
                boolean isBiomeEligible = (!hasType(hotBiome, Type.SAVANNA) && !hasType(hotBiome, Type.JUNGLE));
                if (isBiomeEligible && biome == hotBiome) {
                    this.generateTower(serverworld, random, basePos);
                }
            });
        }
    }

    public void generateTower(WorldServer world, Random random, BlockPos pos) {
        if (tower.enable && ((random.nextInt(tower.chanceNeededForSpawning) < tower.spawnChance) || (enableTowerDevEnv() && (random.nextInt(100) < 10)))) {
            PlacementSettings settings = new PlacementSettings();
            settings.setRotation(Rotation.NONE);
            MinecraftServer server = world.getMinecraftServer();
            Template template = world.getSaveHandler().getStructureTemplateManager().getTemplate(server, TOWER);
            template.addBlocksToWorld(world, pos, settings);
            if (debugMode || enableTowerDevEnv()) {
                LogHelper.info("Tower generated at: " + pos);
            }
        }
    }
}

File: src/main/java/net/thedragonteam/armorplus/worldgen/nbt/StructureGenNBT.kt
/*
 * Copyright (c) TheDragonTeam 2016-2017.
 */

package net.thedragonteam.armorplus.worldgen.nbt

import net.minecraft.init.Blocks
import net.minecraft.util.ResourceLocation
import net.minecraft.util.Rotation
import net.minecraft.util.math.BlockPos
import net.minecraft.world.World
import net.minecraft.world.WorldServer
import net.minecraft.world.chunk.IChunkGenerator
import net.minecraft.world.chunk.IChunkProvider
import net.minecraft.world.gen.structure.template.PlacementSettings
import net.minecraftforge.fml.common.IWorldGenerator
import net.thedragonteam.armorplus.APConfig.*
import net.thedragonteam.armorplus.ArmorPlus
import java.util.*

class StructureGenNBT : IWorldGenerator {

    override fun generate(random: Random, chunkX: Int, chunkZ: Int, world: World, chunkGenerator: IChunkGenerator, chunkProvider: IChunkProvider) {

        val posX = chunkX * 16 + random.nextInt(16)
        val posZ = chunkZ * 16 + random.nextInt(16)

        @Suppress("UNUSED_VARIABLE")
        val posXZ = BlockPos(posX, 1, posZ)

        //        val biome = world.getBiomeForCoordsBody(posXZ)

        if (world !is WorldServer) return

        //        BlockPos playerspawn = serverworld.provider.getSpawnPoint();

        val posY = getGroundFromAbove(world, posX, posZ)
        val basePos = BlockPos(posX, posY, posZ)

        this.generateTower(world, random, basePos)

    }

    fun generateTower(world: WorldServer, random: Random, pos: BlockPos) {
        if (enableTowerGen) {
            if (random.nextInt(towerGenSpawnNeedOfChance) < towerGenSpawnChance) {
                val settings = PlacementSettings()
                settings.rotation = Rotation.NONE

                val server = world.minecraftServer
                val template = world.saveHandler.structureTemplateManager.getTemplate(server, TOWER)

                template.addBlocksToWorld(world, pos, settings)
            }
        }
    }

    /**
     * HELPER METHODS
     */
    // find a grass or dirt block to place the structure on
    fun getGroundFromAbove(world: World, x: Int, z: Int): Int {
        var y = 255
        var foundGround = false
        while (!foundGround && y-- >= 0) {
            val blockAt = world.getBlockState(BlockPos(x, y, z)).block
            // "ground" for our bush is grass or dirt
            foundGround = blockAt === Blocks.DIRT || blockAt === Blocks.GRASS || blockAt !== Blocks.AIR
        }

        return y
    }

    private val TOWER = ResourceLocation(ArmorPlus.MODID, "tower")
}